Task logging handlers can provide custom log links

mdediana opened the pull request:
Airflow provides logging handlers that send logs to external services (Elasticsearch, StackDriver, etc). At the moment, if Elasticsearch is in use, links to its logs show up in the UI. This PR generalizes this functionality to other logging handlers.

To provide links to remote logs, a logging handler must inherit from `ExternalLoggingMixin` and implement `get_remote_log_url()` and `log_name()` methods. The view identifies when a remote logging handler is in use and renders the link to the external logs in the UI.

This PR does not close but is related to issue #9083.    

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

mdediana committed:
Add remote log services redirect

Use a mixin to define log handlers based on remote services. The main
changes are:

 - Create RemoteLoggingMixin to define remote log handlers.

 - Remove explicit mentions to Elasticsearch in dag.html.

 - Rename the /elasticsearch endpoint in views.py to
   /redirect_to_remote_log and dispatch the remote URL building to the
   log handler.

mdediana committed:
Build remote URLs in ES log handler

mdediana committed:
Fix lint issues

mik-laj commented on airflow/utils/log/logging_mixin.py:
```diff
@@ -58,6 +58,20 @@ def _set_context(self, context):
+    Define a logging handler based on a remote service (e.g. ELK, StackDriver). Subclasses must override
+    REMOTE_LOG_NAME and get_remote_log_url().
+    """
+    REMOTE_LOG_NAME: str = 'RemoteLog'
```
```suggestion
    def log_name(self):
        raise NotImplementedError('Remote logger must provide remote name')
```
This will force the user to configure the name.

mik-laj commented on airflow/utils/log/logging_mixin.py:
```diff
@@ -58,6 +58,20 @@ def _set_context(self, context):
             set_context(self.log, context)
 
 
+class RemoteLoggingMixin:
```
```suggestion
class ExternalLoggingMixin:
```
What do you think to change the name? Remote logging is an existing mechanism that allows access to logs from Web UI. In my opinion, external logs better describe this feature.

mdediana replied:
Sounds good, I'll change it.

mik-laj commented on airflow/www/templates/airflow/dag.html:
```diff
@@ -366,9 +366,9 @@ <h4 class="modal-title" id="dagModalLabel">
-    var show_es_logs = false;
-    {% if show_external_logs is defined %}
-      show_es_logs = '{{ show_external_logs }}' == "True";
+    var show_remote_log_redirect = false;
```
```suggestion
    var show_external_log_redirect = false;
```
This name is even more confusing here.

mik-laj commented on airflow/www/views.py:
```diff
@@ -1363,6 +1374,14 @@ def success(self):
                                               future, past, State.SUCCESS)
 
+    @staticmethod
+    def _get_log_handler():
```
Today I was working on one class that contains a similar piece of code. In the future, we will be able to move it.
https://github.com/apache/airflow/pull/9331/commits/530c93c8ff6140e77a3642efd1e9eff3ac6bdd1d

mdediana replied:
Nice. This is also used in `get_logs_with_metadata()`, but I didn't change it because it also uses `task_log_reader`.

mdediana committed:
Rename vars, force log name override

 - Use "external" instead of "remote" in variable names.

 - Force ExternalLoggingMixin subclasses to provide a log_name() method.

mdediana committed:
Fix undefined variable issue

View tests were failing because external_log_name was evaluated even
when show_external_log_redirect is False.

mdediana committed:
Use TaskLogReader

turbaszek commented on airflow/utils/log/es_task_handler.py:
```diff
@@ -262,3 +267,24 @@ def close(self):
 
         self.closed = True
+
+    def log_name(self):
```
```suggestion
    @property
    def log_name(self):
```

turbaszek replied:
WDYT?

mdediana replied:
Done, looks better, thanks. In the process pylint had some complaint about overriden properties, I turned it into an abstract property.

mik-laj commented on airflow/www/views.py:
```diff
@@ -760,22 +761,33 @@ def log(self, session=None):
+            models.TaskInstance.dag_id == dag_id,
+            models.TaskInstance.task_id == task_id,
+            models.TaskInstance.execution_date == dttm).first()
+
```
Can you handle missing TI here? You should probably redirect the user to the main page + display flash message.

mdediana replied:
Done.

mik-laj commented on airflow/www/views.py:
```diff
@@ -760,22 +761,33 @@ def log(self, session=None):
+            error_message = [
+                f"Task log handler does not support external log URLs.\n{str(e)}\n"
+            ]
+            return jsonify(message=error_message, error=True)
```
I am not sure if this is the correct response format. This is not an API, but a view for the end-user, so the response should be text/html.

mdediana replied:
I changed it to redirect to home with an error message, not sure if this is the right thing, let me know.

mik-laj commented on airflow/www/views.py:
```diff
@@ -1493,8 +1505,14 @@ def recurse_nodes(task, visited):
+
+        task_log_reader = TaskLogReader()
+        handler = task_log_reader.log_handler
+        show_external_log_redirect = isinstance(handler, ExternalLoggingMixin)
```
Can you create a property from this condition in TaskLogReader?  In general, all logic should be outside the view method. The view should only format the responses to the expected format. I know that this code does not look like this at the moment, but I am trying to fix it so that the Web UI is not just one file, but it is a more manageable code.

mdediana replied:
Done.

mdediana committed:
Address PR comments

The changes are:

 - Handle missing TI in "views.redirect_to_external_log()".

 - Redirect to home and show an error message if try to redirect to
   external log when using a local log handler.

 - Create "TaskLogReader.is_external()".

mdediana committed:
Make external log name a property

Convert ExternalLoggingMixin.log_name() from method into property.
Also make ExternalLoggingMixin an abstract class instead of relying on
NotImplementedError.

mik-laj commented on airflow/utils/log/log_reader.py:
```diff
@@ -100,6 +101,11 @@ def is_supported(self):
 
+    @property
+    def is_external(self):
+        """Check if the logging handler is external (e.g. Elasticsearch, Stackdriver, etc)."""
```
```suggestion
    def supports_external_link(self):
        """Check if the logging handler is external links (e.g. Elasticsearch, Stackdriver, etc)."""
```
This name is because the word "external" suggests that the class is not part of the Airflow core, but we want to check here whether external links are supported.  When we introduce it, we need to improve the "is_supported" property so that it has a similar naming convention - `support_read`

mdediana replied:
Makes sense.

mik-laj commented on airflow/www/views.py:
```diff
@@ -760,22 +760,35 @@ def log(self, session=None):
+            return redirect(url)
+        except AttributeError:
+            flash("Cannot redirect to external log, task log handler is not external", "error")
+            return redirect(url_for('Airflow.index'))
```
```suggestion
        task_log_reader = TaskLogReader()
        if not task_log_reader.supports_external_links:
            flash("Ttask log handler is not support external links", "error")
            return redirect(url_for('Airflow.index'))

        handler = task_log_reader.log_handler
        url = handler.get_external_log_url(ti, try_number)
        return redirect(url)
```
This exception here may mean that the handler does not support external links or that a failure occurred while generating links. We have another easy way to check if an error occurs, so we can use it.

mdediana committed:
Use default exception handling

Co-authored-by: Kamil Breguła <mik-laj@users.noreply.github.com>

mdediana committed:
Rename property

mdediana committed:
Add documentation for external links

mdediana committed:
Fix ES remote logging doc

mik-laj commented on airflow/utils/log/log_reader.py:
```diff
@@ -100,6 +101,11 @@ def is_supported(self):
 
```
Can you rename `is_supported` to `supports_read` or somethings similar?

mik-laj commented on airflow/utils/log/es_task_handler.py:
```diff
@@ -262,3 +267,25 @@ def close(self):
+    def get_external_log_url(self, task_instance: TaskInstance, try_number: int) -> str:
+        """
+        Creates an address for an external log collecting service.
+        :param task_instance: task instance object
```
```suggestion

        :param task_instance: task instance object
```
Without this, the documentation does not comply with the Sphinx.

mik-laj reviewed (commented):
We are very close to the end. 
More people on Slack are asking about this feature.
https://apache-airflow.slack.com/archives/CCRR5EBA7/p1593356282393500

mdediana committed:
Change doc to comply with Sphinx

Co-authored-by: Kamil Breguła <mik-laj@users.noreply.github.com>

mdediana committed:
Rename property

turbaszek reviewed (approved):
LGTM!
